controllers/policystatus_utils.go

```diff
@@ -106,12 +115,11 @@ func isPolicyUniquelyReachable(ctx context.Context, apiReader client.Reader, pol
 		if pod.DeletionTimestamp != nil {
 			continue
 		}
-		if pod.Labels[appsv1.DefaultDeploymentUniqueLabelKey] != podTemplateHash || !isPodReady(pod) {
-			return false
+		if pod.Labels[appsv1.DefaultDeploymentUniqueLabelKey] == podTemplateHash && isPodReady(pod) {
+			return true
```

Comment:
This is inverting the logic of the check. The idea of the `isUniquelyReachable` condition is that there is no possibility for a request to arrive to an old pod (e.g. with old policy settings, or old policy version), because no old pod exists.

So, the fact that we find a pod coming from the newest replicaset and that it's ready does not mean that other older pod from an older replicaset is not still around able to process requests.

I would change this code to follow this behavior. Furthermore, I think we can also change the policy status before:

- Move lines https://github.com/kubewarden/kubewarden-controller/blob/410204ee09fdcfa50f78296d79c8c887e7b3e508/controllers/policy_utils.go#L107-L118 to https://github.com/kubewarden/kubewarden-controller/blob/410204ee09fdcfa50f78296d79c8c887e7b3e508/controllers/policy_utils.go#L133.
-  Move https://github.com/kubewarden/kubewarden-controller/blob/410204ee09fdcfa50f78296d79c8c887e7b3e508/controllers/policy_utils.go#L143 right before the `isUniquelyReachable` block just moved in the previous point.

This would ensure that the policy is set to active soon, but that we only change the `isUniquelyRechable` condition at the very end, when we know that no old pods will receive any requests. Users are free to wait for the isUniquelyReachable condition or not. When waiting for it we know that no request can reach an older pod, but we also know that this condition takes a little more to transition to `True`.

So, what I propose is something like:

```go
func reconcilePolicy(ctx context.Context, client client.Client, reconciler admission.Reconciler, policy v1alpha2.Policy) (ctrl.Result, error) {
	if policy.GetPolicyServer() == "" {
		policy.SetStatus(v1alpha2.PolicyStatusUnscheduled)
		return ctrl.Result{}, nil
	}
	policyServer, err := getPolicyServer(ctx, client, policy)
	if err != nil {
		policy.SetStatus(v1alpha2.PolicyStatusScheduled)
		//lint:ignore nilerr set status to scheduled if policyServer can't be retrieved, and stop reconciling
		return ctrl.Result{}, nil
	}
	if policy.GetStatus().PolicyStatus != v1alpha2.PolicyStatusActive {
		policy.SetStatus(v1alpha2.PolicyStatusPending)
	}

	policyServerDeployment := appsv1.Deployment{}
	if err := reconciler.APIReader.Get(ctx, types.NamespacedName{Namespace: reconciler.DeploymentsNamespace, Name: naming.PolicyServerDeploymentNameForPolicyServerName(policy.GetPolicyServer())}, &policyServerDeployment); err != nil {
		if apierrors.IsNotFound(err) {
			return ctrl.Result{Requeue: true}, nil
		}
		return ctrl.Result{}, errors.Wrap(err, "could not read policy server Deployment")
	}

	secret := corev1.Secret{}
	if err := client.Get(ctx, types.NamespacedName{Namespace: reconciler.DeploymentsNamespace, Name: constants.PolicyServerCARootSecretName}, &secret); err != nil {
		return ctrl.Result{}, errors.Wrap(err, "cannot find policy server secret")
	}
	if policy.IsMutating() {
		if err := reconciler.ReconcileMutatingWebhookConfiguration(ctx, policy, &secret, policyServer.NameWithPrefix()); err != nil {
			return ctrl.Result{}, errors.Wrap(err, "error reconciling mutating webhook")
		}
	} else {
		if err := reconciler.ReconcileValidatingWebhookConfiguration(ctx, policy, &secret, policyServer.NameWithPrefix()); err != nil {
			return ctrl.Result{}, errors.Wrap(err, "error reconciling validating webhook")
		}
	}

	policy.SetStatus(v1alpha2.PolicyStatusActive)

	if !isPolicyUniquelyReachable(ctx, client, &policyServerDeployment) {
		apimeta.SetStatusCondition(
			&policy.GetStatus().Conditions,
			metav1.Condition{
				Type:    string(v1alpha2.PolicyUniquelyReachable),
				Status:  metav1.ConditionFalse,
				Reason:  "LatestReplicaSetIsNotUniquelyReachable",
				Message: "The latest replica set is not uniquely reachable",
			},
		)
		return ctrl.Result{Requeue: true, RequeueAfter: 2 * time.Second}, nil
	}

	apimeta.SetStatusCondition(
		&policy.GetStatus().Conditions,
		metav1.Condition{
			Type:    string(v1alpha2.PolicyUniquelyReachable),
			Status:  metav1.ConditionTrue,
			Reason:  "LatestReplicaSetIsUniquelyReachable",
			Message: "The latest replica set is uniquely reachable",
		},
	)

	return ctrl.Result{}, nil
}
```

Reply:
This change break the tests. But I'll find a solution for that and push the code for review again.

Reply:
I've pushed a new version of the `isPolicyUniquelyReachable` function. Now, I do not invert the logic of it. But I filter the replicaset used in the comparison. AFAICS, it fixed the original problem and it does not require changes in the tests. 

Let me know what you think